crates/matrix-sdk-indexeddb/src/stream.rs

```diff
@@ -0,0 +1,624 @@
+                results,
+                Ok(vec![
+                    vec![JsValue::from_f64(0.), JsValue::from_f64(1.),],
+                    vec![JsValue::from_f64(2.), JsValue::from_f64(3.),],
+                    vec![JsValue::from_f64(4.),]
+                ])
+            );
+        }
+
+        Ok(())
+    }
+
+    #[wasm_bindgen_test]
+    async fn stream_by_renewed_cursor() -> Result<(), DomException> {
+        let mut db = IdbDatabase::open("bazqux")?;
+
+        db.set_on_upgrade_needed(Some(|event: &IdbVersionChangeEvent| -> Result<(), JsValue> {
+            if event.db().object_store_names().find(|n| n == "baz").is_none() {
+                event.db().create_object_store("baz")?;
+            }
+
+            Ok(())
+        }));
+
+        let db = db.await?;
+
+        {
+            let transaction =
+                db.transaction_on_one_with_mode("baz", IdbTransactionMode::Readwrite)?;
+            let object_store = transaction.object_store("baz")?;
+
+            for nth in 0..5 {
+                object_store.add_key_val_owned(
+                    JsValue::from_str(&format!("key{nth}")),
+                    &JsValue::from(nth),
+                )?;
+            }
+
+            transaction.await.into_result()?;
+        }
+
+        // Test one read. Nothing particular here.
+        {
+            let transaction = db.transaction_on_one("baz")?;
+            let store = transaction.object_store("baz")?;
+
+            let value: Option<JsValue> = store.get_owned("key3")?.await?;
+            assert_eq!(value.expect("value is none").as_f64().expect("boo"), 3.);
+        }
+
+        // Stream!
+        {
+            let mut number_of_renews = 0;
+            let mut stream = StreamByRenewedCursor::new(
+                || {
+                    number_of_renews += 1;
+                    Ok(db.transaction_on_one("baz")?)
+                },
+                "baz".to_owned(),
+                // SAFETY: `unwrap` is safe because 3 is not zero.
+                NonZeroUsize::new(3).unwrap(),
+            )
+            .await?;
+
+            assert_eq!(
+                stream.try_next().await?,
+                Some((JsValue::from_str("key0"), JsValue::from_f64(0.)))
+            );
+            assert_eq!(
+                stream.try_next().await?,
+                Some((JsValue::from_str("key1"), JsValue::from_f64(1.)))
+            );
+            assert_eq!(
+                stream.try_next().await?,
+                Some((JsValue::from_str("key2"), JsValue::from_f64(2.)))
+            );
+            assert_eq!(
+                stream.try_next().await?,
+                Some((JsValue::from_str("key3"), JsValue::from_f64(3.)))
+            );
+            assert_eq!(
+                stream.try_next().await?,
+                Some((JsValue::from_str("key4"), JsValue::from_f64(4.)))
+            );
+            assert_eq!(stream.try_next().await?, None);
+
+            assert_eq!(number_of_renews, 2);
```

Comment:
I would like 2 more assertions on `number_of_renews`: when it is zero, and when it is 1.